File: CLASE VII - Listas Void con Struct Anidadas/pre_parcial/c/lista.c
#include <stdio.h>
#include <stdlib.h>

#include "lista.h"

struct Lista {
    NodoPtr primero;
};

ListaPtr crearLista() {
    ListaPtr lista = (ListaPtr) malloc(sizeof(struct Lista));
    lista->primero = NULL;

    return lista;
}

NodoPtr getPrimero(ListaPtr lista) {
    return lista->primero;
}
void setPrimero(ListaPtr lista, NodoPtr nuevoPrimero) {
    lista->primero = nuevoPrimero;
}

// Insertar:
void insertarPrimero(ListaPtr lista, DatoPtr dato) {
    NodoPtr nuevoNodo = crearNodo(dato, lista->primero);
    lista->primero = nuevoNodo;
}
void insertarUltimo(ListaPtr lista, DatoPtr dato) {
    NodoPtr nuevoNodo = crearNodo(dato, NULL);


    if(lista->primero == NULL) {
        lista->primero = nuevoNodo;
    } else {
        NodoPtr actual = lista->primero;
        while(getSiguiente(actual)!=NULL) {
            actual = getSiguiente(actual);
        }
        setSiguiente(actual, nuevoNodo);
    }
}
void insertarPosicion(ListaPtr lista, DatoPtr dato, int posicion) {
    int tam = obtenerTamanio(lista);
    if(posicion > tam+1) {
        printf("No existe esa posicion\n");
    }
    else if(posicion < 0) {
        printf("No existe esa posicion\n");
    }
    else if(posicion == 0) {
        insertarPrimero(lista, dato);
    }
    else if(posicion == tam) {
        insertarUltimo(lista, dato);
    }
    else if(posicion > 0 && posicion < tam){
        NodoPtr actual = lista->primero;
        for(int i=0; i<posicion-1; i++){
            actual = getSiguiente(actual);
        }
        setSiguiente(actual, crearNodo(dato, getSiguiente(actual)));
    }
}

// Eliminar:
DatoPtr eliminarPrimero(ListaPtr lista) {
    DatoPtr dato = NULL;
    NodoPtr actual = lista->primero;

    if(actual != NULL) {
        dato = getDato(actual);
        NodoPtr actual = lista->primero;
        lista->primero = getSiguiente(actual);
        liberarNodo(actual);
    }

    return dato;
}
DatoPtr eliminarPosicion(ListaPtr lista, int pos) {
    NodoPtr actual = lista->primero;
    NodoPtr anterior = NULL;
    DatoPtr dato = NULL;
    int tam = obtenerTamanio(lista);
    int i=0;

    if(actual == NULL){
        printf("Lista vacia\n");
    }
    else if (pos < 0 || pos >= tam) {
        printf("Posicion fuera de rango\n");
    }
    else if(pos == 0){
        dato = eliminarPrimero(lista);
    }
    else if(pos > 0 || pos < tam) {    //pos dentro de la lista
        while(i != pos) {
            anterior = actual;
            actual = getSiguiente(actual);
            i++;
        }
        dato = getDato(actual);
        setSiguiente(anterior, getSiguiente(actual));
        liberarNodo(actual);
    }
    return dato;
}

// Obtener:
NodoPtr obtenerPrimero(ListaPtr lista) {
    return lista->primero;
}
int obtenerTamanio(ListaPtr lista) {
    int tam = 0 ;
    NodoPtr actual = lista->primero;
    while (actual != NULL) {
        tam = tam + 1;
        actual = getSiguiente(actual);
    }

    return tam;
}



void mostrarListaGenerica(ListaPtr lista, void(*mostrar)(DatoPtr)){

    printf("\n<LISTA> TAM:%d \n", obtenerTamanio(lista));

    NodoPtr actual = lista->primero;

    while (actual!=NULL){

        mostrar(getDato(actual));
        actual = getSiguiente(actual);
    }
    printf("\n\n");



};



// Mostrar/Liberar/Duplicar:
void mostrarLista(ListaPtr lista, void (*mostrarDato)(NodoPtr)) {
    //printf("<LISTA> TAM:%d \n", obtenerTamanio(lista));
    NodoPtr actual = lista->primero;

    while (actual!=NULL) {
        mostrarDato(actual);
        actual = getSiguiente(actual);
    }
    printf("\n");
};
void liberarLista(ListaPtr lista) {
    NodoPtr actual = lista->primero;
    NodoPtr aux;

    while (actual!=NULL) {
        aux = actual;
        actual = getSiguiente(actual);
        free(aux);
    };
}

File: CLASE VII - Listas Void con Struct Anidadas/pre_parcial/golang/lista.go
package main

import "fmt"

type Nodo[T any] struct {
	dato      T
	siguiente *Nodo[T]
}

func crearNodo[T any](dato T, siguiente *Nodo[T]) *Nodo[T] {
	return &Nodo[T]{dato: dato, siguiente: siguiente}
}

type Lista[T any] struct {
	primero *Nodo[T]
}

func crearLista[T any]() *Lista[T] {
	return &Lista[T]{primero: nil}
}

func getPrimero[T any](lista *Lista[T]) *Nodo[T] {
	return lista.primero
}

func setPrimero[T any](lista *Lista[T], nuevoPrimero *Nodo[T]) {
	lista.primero = nuevoPrimero
}

func insertarPrimero[T any](lista *Lista[T], dato T) {
	lista.primero = crearNodo(dato, lista.primero)
}

func insertarUltimo[T any](lista *Lista[T], dato T) {
	nuevo := crearNodo(dato, nil)
	if lista.primero == nil {
		lista.primero = nuevo
		return
	}
	actual := lista.primero
	for actual.siguiente != nil {
		actual = actual.siguiente
	}
	actual.siguiente = nuevo
}

func obtenerTamanio[T any](lista *Lista[T]) int {
	tam := 0
	actual := lista.primero
	for actual != nil {
		tam++
		actual = actual.siguiente
	}
	return tam
}

func eliminarPrimero[T any](lista *Lista[T]) (dato T, ok bool) {
	if lista.primero == nil {
		var zero T
		return zero, false
	}
	dato = lista.primero.dato
	lista.primero = lista.primero.siguiente
	return dato, true
}

func eliminarPosicion[T any](lista *Lista[T], pos int) (dato T, ok bool) {
	tam := obtenerTamanio(lista)
	if lista.primero == nil {
		fmt.Printf("Lista vacia\n")
		var zero T
		return zero, false
	}
	if pos < 0 || pos >= tam {
		fmt.Printf("Posicion fuera de rango\n")
		var zero T
		return zero, false
	}
	if pos == 0 {
		return eliminarPrimero(lista)
	}

	anterior := (*Nodo[T])(nil)
	actual := lista.primero
	i := 0
	for i != pos {
		anterior = actual
		actual = actual.siguiente
		i++
	}
	dato = actual.dato
	anterior.siguiente = actual.siguiente
	return dato, true
}

func mostrarListaGenerica[T any](lista *Lista[T], mostrar func(T)) {
	fmt.Printf("\n<LISTA> TAM:%d \n", obtenerTamanio(lista))
	actual := lista.primero
	for actual != nil {
		mostrar(actual.dato)
		actual = actual.siguiente
	}
	fmt.Printf("\n\n")
}

